include/nbl/builtin/hlsl/spirv_intrinsics/core.hlsl

```diff
@@ -47,199 +69,1272 @@ static const uint32_t3 LocalInvocationId;
+T genericPtrMemSemantics(pointer_t<spv::StorageClassPhysicalStorageBuffer, T> pointer);
+
+template<typename T>
+[[vk::ext_capability(spv::CapabilityBitInstructions)]]
+[[vk::ext_instruction(spv::OpBitFieldInsert)]]
+T bitFieldInsert(T base, T insert, uint32_t offset, uint32_t count);
+
+[[vk::ext_capability(spv::CapabilityBitInstructions)]]
+[[vk::ext_instruction(spv::OpBitFieldSExtract)]]
+int32_t bitFieldExtract(int32_t base, uint32_t offset, uint32_t count);
+
+[[vk::ext_capability(spv::CapabilityBitInstructions)]]
+[[vk::ext_instruction(spv::OpBitFieldSExtract)]]
+int64_t bitFieldExtract(int64_t base, uint32_t offset, uint32_t count);
+
+[[vk::ext_capability(spv::CapabilityBitInstructions)]]
+[[vk::ext_instruction(spv::OpBitFieldUExtract)]]
+uint32_t bitFieldExtract(uint32_t base, uint32_t offset, uint32_t count);
+
+[[vk::ext_capability(spv::CapabilityBitInstructions)]]
+[[vk::ext_instruction(spv::OpBitFieldUExtract)]]
+uint64_t bitFieldExtract(uint64_t base, uint32_t offset, uint32_t count);
+
+template<typename T>
+[[vk::ext_capability(spv::CapabilityBitInstructions)]]
+[[vk::ext_instruction(spv::OpBitReverse)]]
+T bitReverse(T base);
+
+template<typename T>
+[[vk::ext_instruction(spv::OpBitCount)]]
+T bitCount(T base);
+
+[[vk::ext_instruction(spv::OpControlBarrier)]]
+void controlBarrier(uint32_t executionScope, uint32_t memoryScope,  uint32_t semantics);
+
+[[vk::ext_instruction(spv::OpMemoryBarrier)]]
+void memoryBarrier(uint32_t memoryScope,  uint32_t semantics);
+
+template<typename T>
+[[vk::ext_instruction(spv::OpAtomicLoad)]]
+T atomicLoad([[vk::ext_reference]] T pointer, uint32_t memoryScope,  uint32_t semantics);
+
+template<typename T, typename P>
+[[vk::ext_instruction(spv::OpAtomicLoad)]]
+enable_if_t<is_spirv_type_v<P>, T> atomicLoad(P pointer, uint32_t memoryScope,  uint32_t semantics);
+
+template<typename T>
+[[vk::ext_instruction(spv::OpAtomicStore)]]
+void atomicStore([[vk::ext_reference]] T pointer, uint32_t memoryScope,  uint32_t semantics, T value);
+
+template<typename T, typename P>
+[[vk::ext_instruction(spv::OpAtomicStore)]]
+enable_if_t<is_spirv_type_v<P>, void> atomicStore(P pointer, uint32_t memoryScope,  uint32_t semantics, T value);
+
+template<typename T>
+[[vk::ext_instruction(spv::OpAtomicExchange)]]
+T atomicExchange([[vk::ext_reference]] T pointer, uint32_t memoryScope,  uint32_t semantics, T value);
+
+template<typename T, typename P>
+[[vk::ext_instruction(spv::OpAtomicExchange)]]
+enable_if_t<is_spirv_type_v<P>, T> atomicExchange(P pointer, uint32_t memoryScope,  uint32_t semantics, T value);
 
-template<typename T> // integers operate on 2s complement so same op for signed and unsigned
-[[vk::ext_capability(spv::CapabilityInt64Atomics)]]
+template<typename T>
+[[vk::ext_instruction(spv::OpAtomicCompareExchange)]]
+T atomicCompareExchange([[vk::ext_reference]] T pointer, uint32_t memoryScope,  uint32_t equal,  uint32_t unequal, T value, T comparator);
```

Comment:
btw does the SPIR-V spec say that any of the memory operands need to be literals or not?

Reply:
No, just checked